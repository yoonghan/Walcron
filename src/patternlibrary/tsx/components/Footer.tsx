`use strict`

import * as React from "react";
import { LocaleSelector, LocaleSelectorTypes } from "./LocaleSelector";
import { UtilLocale } from "../util/UtilLocale";
import * as CSSTransition from 'react-transition-group/CSSTransition';

import '../../scss/base';
var styles = require('../../scss/components/Footer');
declare function require(path: string): any;

interface ContactOpenState {
  isContactOpen: boolean;
  contactIcon: string;
}

export interface LinksItem {
  link: string;
  label: string;
}

export interface FooterProps {
  links: Array<LinksItem>;
  updatedYear: number;
  emailLabel: string;
  emailTitle: string;
  emailAddress: string;
  emailMsg: string;
}

export class Footer extends React.Component<FooterProps, ContactOpenState> {

  private node: HTMLElement;

  constructor(props:any) {
    super(props);

    const openState = false;

    this.state = {
      isContactOpen: openState,
      contactIcon: this.displayContactIcon(openState)
    };
  };

  clickContact = () => {
    const isContactOpen = !this.state.isContactOpen,
          self = this;

    this.setState(
      (prevState, props) => {
        return {
          isContactOpen: isContactOpen,
          contactIcon: self.displayContactIcon(isContactOpen)
        };
      }
    );
  };

  componentDidUpdate() {
    if(this.state.isContactOpen) {
      const scrollDuration = 500;
      const self = this.node;
      const pos = self.getBoundingClientRect().height;

      if(pos > 0) {
        var scrollHeight = pos;
        const scrollStep = Math.PI / ( scrollDuration / 15 ),
            cosParameter = scrollHeight / 2;
        var scrollCount = 0,
            scrollMargin,
            scrollInterval = setInterval( function() {
              if ( scrollHeight > 0 ) {
                  scrollCount = scrollCount + 1;
                  scrollMargin = cosParameter - cosParameter * Math.cos( scrollCount * scrollStep );
                  window.scrollTo( 0, ( window.scrollY + scrollMargin ) );
                  scrollHeight -= scrollMargin;
              }
              else clearInterval(scrollInterval);
            }, 15 );
      }
    }
  };

  displayContactIcon = (isShown:boolean):string => {
    return (isShown ? 'remove' : 'phone');
  };

  loadContactInfo = () => {
    const {emailAddress, emailTitle, emailMsg} = this.props;

    return (
      <div ref={node => this.node = node} className={styles['footer-contact']}>
        <h4>{emailTitle}</h4>
        <div>
          <a href={'mailto:' + emailAddress}>
            <i className='fa fa-pencil-square-o'></i> {emailAddress}
          </a>
        </div>
        <div>{emailMsg}</div>
      </div>
    );
  }

  render() {
    const {links, updatedYear, emailLabel} = this.props;

    const linkList = links.map(function(item) {
      const {link, label} = item;
      return (
        <a href={UtilLocale.getLocalizedHref(link)} key={label}>{label}</a>
      );
    });

    return (
      <div className={styles.footer + ' links-white-background'}>
        <div className={styles['footer-divider']}/>
        <div className={styles['footer-container']}>
          <div className={styles['footer-section-one']}>
            <a href={UtilLocale.getLocalizedHref('/')}>
              <img src="/ext/img/logo/logoOnWhiteBg.svg"/>
            </a>
            <div>
              <LocaleSelector type={LocaleSelectorTypes.Dropdown}/>
            </div>
            <div className={styles['footer-section-one-link']}>
              <a href="javascript:;" onClick={this.clickContact} className='links'>
                <i className={'fa fa-' + this.state.contactIcon}></i>
                {!this.state.isContactOpen && emailLabel}
              </a>
            </div>
          </div>
          <CSSTransition
            in = {
              this.state.isContactOpen
            }
            classNames={{
              enter: styles['footercontactemail-enter'],
              enterActive: styles['footercontactemail-enter-active'],
              exit: styles['footercontactemail-leave'],
              exitActive: styles['footercontactemail-leave-active']
            }}
            timeout={{
              enter: 300,
              exit: 300
            }}>
            <div>
              {this.state.isContactOpen && this.loadContactInfo()}
            </div>
          </CSSTransition>
          <hr/>
          <div className={styles['footer-section-two'] + ' links-white-background'}>
            {linkList}
          </div>
          <div className={styles['footer-section-three'] + ' base-footnote'}>
            2016-{updatedYear} Walcron Coorperation &copy; 
          </div>
        </div>
      </div>
    );
  }
}
