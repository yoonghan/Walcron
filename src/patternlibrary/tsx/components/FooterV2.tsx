`use strict`

import * as React from "react";
import * as CSSTransition from 'react-transition-group/CSSTransition';
import {MenuIcon} from './MenuIcon';

import '../../scss/base';
var styles = require('../../scss/components/FooterV2');

interface ContactOpenState {
  isContactOpen: boolean;
}

interface ILinks {
  text: string;
  path: string;
}

interface ContactOpenProps {
  contactInformation: string;
}

export interface FooterProps {
  updatedYear: number;
  contactInformation: string;
  linkArray: Array<ILinks>;
  isHomepage?: boolean;
}

export class FooterV2 extends React.Component<FooterProps, {}> {
  constructor(props:any) {
    super(props);
  }

  _isOwnPage = (pathToCheck:string) => {
    const pathName = location.pathname;
    if(pathName.indexOf('/') === -1 || pathToCheck === '/') {
      return false;
    }

    const currentPageIdx = pathName.lastIndexOf('/'),
      path = pathName.substring(currentPageIdx);

    return path.indexOf(pathToCheck) === 0;
  }

  _generateLinks = (links:Array<ILinks>) => {
    const linkAsHtml = links.map(
      (linkItem) => {
        const {text, path} =  linkItem;

        if(this._isOwnPage(path)) {
          return;
        }

        return (
           <a href={path} key={text}>&bull; {text}</a>
        );
      }
    );

    if(!this.props.isHomepage) {
      linkAsHtml.unshift(<a href="/" className={styles['ftr-back-arrow']}><i className={"fa fa-arrow-circle-left"}></i></a>);
    }

    return linkAsHtml;
  };

  render() {
    const {updatedYear, linkArray, contactInformation} = this.props;
    const generatedLinks = this._generateLinks(linkArray);

    return (
      <div className={styles.ftr}>
        <div className={styles['ftr-container']}>
          <div className={styles['ftr-section-one']}>
            <ContactDisplay contactInformation={contactInformation}/>
          </div>
          <div className={styles['ftr-divider']}></div>
          <div className={styles['ftr-links']}>
            {generatedLinks}
          </div>
          <div className={styles['ftr-divider']}></div>
          <div className={styles['ftr-footnote']}>
            2016 - {updatedYear} Walcron Coorperation &copy;
          </div>
        </div>
      </div>
      );
  }
}

class ContactDisplay extends React.Component<ContactOpenProps, ContactOpenState> {
  constructor(props:any) {
    super(props);

    const openState = false;

    this.state = {
      isContactOpen: openState
    };
  }

  _generateContactOpen = () => {
    return (
      <div className={styles['ftr-contact-close-container']}>
        <div className={styles['ftr-contact-email-content']}>
          <div className={styles['ftr-contact-close-deco']}>
            <a href="javascript:;" className={styles['ftr-contact-close-btn'] + ' links'} onClick={this._clickContact}>
              <i className={"fa fa-remove"}></i>
            </a>
          </div>
          <div className={styles['ftr-contact-email-information']}>
            {this.props.contactInformation}
          </div>
          <div className={styles['ftr-contact-email-contact']}>
            <a href="mailto:walcoorperation@gmail.com">
              <i className="fa fa-envelope-open"></i>
            </a>
            <a href="https://github.com/yoonghan" target="contact">
              <i className="fa fa-github"></i>
            </a>
            <a href="https://www.facebook.com/walcron.coorperation" target="contact">
              <i className="fa fa-facebook-official"></i>
            </a>
          </div>
        </div>
      </div>
    );
  }

  _generateNonContactOpen = () => {
    return (
      <div className={styles['ftr-contact-open-deco']}>
        <a href="javascript:;" className={styles['ftr-contact-open-btn'] + ' links'} onClick={this._clickContact}>
          <span>
            <i className={'fa fa-pencil-square-o'}></i>
            <span>
              Contact Us
            </span>
          </span>
        </a>
      </div>
    );
  }

  _clickContact = () => {

    window.scrollTo(0, document.body.scrollHeight || document.documentElement.scrollHeight);

    this.setState(
      (prevState, props) => {
        return {
          isContactOpen: !prevState.isContactOpen
        };
      }
    );
  }

  render() {
    const {isContactOpen} = this.state;
    const generatedContactOpen = this._generateContactOpen(),
      generatedNonContactOpen = this._generateNonContactOpen();

    return (
      <div className={styles['ftr-contact']}>
        <CSSTransition
          in = {isContactOpen}
          classNames={{
            enter: styles['footercontactemail-enter'],
            enterActive: styles['footercontactemail-enter-active'],
            exit: styles['footercontactemail-leave'],
            exitActive: styles['footercontactemail-leave-active']
          }}
          timeout={{
            enter: 500,
            exit: 500
          }}>
            <div>{isContactOpen && generatedContactOpen}</div>
        </CSSTransition>
        {!isContactOpen && generatedNonContactOpen}
      </div>
    );
  }
}
