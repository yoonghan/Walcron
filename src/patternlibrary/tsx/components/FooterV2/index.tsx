`use strict`

import * as React from "react";
import * as CSSTransition from 'react-transition-group/CSSTransition';
import {MenuIcon} from '../MenuIcon';
import {ContactDisplay} from './ContactDisplay';

import '../../../scss/base.scss';
var styles = require('../../../scss/components/FooterV2.scss');

interface ILinks {
  text: string;
  path: string;
}

export interface FooterProps {
  updatedYear: number;
  contactInformation: string;
  linkArray: Array<ILinks>;
  isHomepage?: boolean;
}

export class FooterV2 extends React.PureComponent<FooterProps, {}> {
  constructor(props:any) {
    super(props);
  }

  _isOwnPage = (pathToCheck:string) => {
    const pathName = location.pathname;
    if(pathName.indexOf('/') === -1 || pathToCheck === '/') {
      return false;
    }

    const currentPageIdx = pathName.lastIndexOf('/'),
      path = pathName.substring(currentPageIdx);

    return path.indexOf(pathToCheck) === 0;
  }

  _generateLinks = (links:Array<ILinks>) => {
    const linkAsHtml = links.map(
      (linkItem) => {
        const {text, path} =  linkItem;

        if(this._isOwnPage(path)) {
          return;
        }

        return (
           <span>&bull; <a href={path} key={text}>{text}</a></span>
        );
      }
    );

    if(!this.props.isHomepage) {
      linkAsHtml.unshift(<a href="/" className={styles['ftr-back-arrow']}><i className={"fa fa-arrow-circle-left"}></i></a>);
    }

    return linkAsHtml;
  };

  render() {
    const {updatedYear, linkArray, contactInformation} = this.props;
    const generatedLinks = this._generateLinks(linkArray);

    return (
      <div className={styles.ftr}>
        <div className={styles['ftr-container']}>
          <div className={styles['ftr-section-one']}>
            <ContactDisplay contactInformation={contactInformation}/>
          </div>
          <div className={styles['ftr-divider']}></div>
          <div className={styles['ftr-links']}>
            {generatedLinks}
          </div>
          <div className={styles['ftr-divider']}></div>
          <div className={styles['ftr-footnote']}>
            2014 - {updatedYear} Walcron Coorperation &copy;
          </div>
        </div>
      </div>
      );
  }
}
