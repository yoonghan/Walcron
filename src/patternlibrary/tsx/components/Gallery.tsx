`use strict`

/**
*Rotating image background for Overlay Indicator is saved as based 64 in scss
**/

import * as React from "react";
import {Button} from 'react-toolbox/lib/button';
import {Transition} from 'react-transition-group';

import '../../scss/base';
var styles = require('../../scss/components/Gallery');
declare function require(path: string): any;

interface GalleryState {
  isOverlayOpen: boolean;
  clickPosLeft: number;
  clickPosTop: number;
  selectedIdx: number;
}

interface MoziacProps {
  imgSrc: string;
  title: string;
  idx: number;
  clickHandler:any;
}

interface OverlayProps {
  isSubComponentToUpdate: boolean;
  transitionStyles: any;
  clickHandler:any;
  animLeft: number;
  animTop: number;
  items: Array<Item>;
  selectedIdx: number;
}

interface OverlayCardProps {
  isComponentUpdated: boolean;
  clickHandler:any;
  animLeft: number;
  animTop: number;
  items: Array<Item>;
  selectedIdx: number;
}

interface OverlayCardAnimatedProps {
  imgSrc: string;
}

interface OverlayIndicatorProps {
  total: number;
  selection: number;
}

export interface GalleryProps {
  items: Array<Item>;
}

export interface Item {
  title: string;
  imgSrc: string;
  nature: Array<string>;
  key: string;
  desc: string;
}

export class Gallery extends React.Component<GalleryProps, GalleryState> {

  private transitionStyles:any = {
    entering: {opacity:0},
    entered: {opacity:1},
    exiting: {opacity: 0, transition: 'opacity ' + Overlay.leaveTime + 'ms ease-in'}
  };

  constructor(props:any) {
    super(props);
    this.state = {
      isOverlayOpen: false,
      clickPosLeft: 0,
      clickPosTop: 0,
      selectedIdx: 0
    };
  };

  createMoziacListing = ():JSX.Element[] => {
    const moziacListing = this.props.items;

    let idx = 0;
    return moziacListing.map(
      (moziac) =>{
        return (
          <Moziac clickHandler={this.toggleOverlay} imgSrc={moziac.imgSrc} title={moziac.title} key={moziac.title} idx={idx++}></Moziac>
        );
      }
    );
  }

  toggleOverlay = (posTop:number = 0, posLeft:number = 0, selectedIdx:number = 0) => {
    const {isOverlayOpen} = this.state;

    this.setState(
      (prevState, props) => {
        return {
          isOverlayOpen: !isOverlayOpen,
          clickPosLeft: posLeft,
          clickPosTop: posTop,
          selectedIdx: selectedIdx
        };
      }
    );
  };

  render() {



    return (
      <div className={styles.gallery}>
            <Transition in={this.state.isOverlayOpen} timeout={Overlay.leaveTime} appear={this.state.isOverlayOpen}>
            {
              (state:any) =>
                {
                  if(state === 'exited') {
                    return null;
                  }

                  return (
                      <Overlay
                        transitionStyles={...this.transitionStyles[state]}
                        clickHandler={this.toggleOverlay}
                        animTop={this.state.clickPosTop}
                        animLeft={this.state.clickPosLeft}
                        items={this.props.items}
                        selectedIdx={this.state.selectedIdx}
                        isSubComponentToUpdate={state==='entering'}/>
                  );
                }
            }
            </Transition>
        {this.createMoziacListing()}
      </div>
    );
  }
}

class Overlay extends React.Component<OverlayProps, {}> {
  private node:HTMLElement;
  static leaveTime = 100;

  _handleClickHandler = () => {
    this.props.clickHandler();
  };

  shouldComponentUpdate() {
    return true;
  }

  render() {
    const {animLeft, animTop, items, selectedIdx, isSubComponentToUpdate} = this.props;
    return (
      <div ref={node => this.node = node} className={styles.overlay} style={{...this.props.transitionStyles}}>
        <div className={styles['overlay-background']} onClick={this._handleClickHandler}></div>
        <OverlayCard
          animLeft={animLeft}
          animTop={animTop}
          clickHandler={this._handleClickHandler}
          items={items}
          selectedIdx={selectedIdx}
          isComponentUpdated={isSubComponentToUpdate}/>
      </div>
    );
  }
}

class OverlayCard extends React.Component<OverlayCardProps, {}> {
  private node:HTMLElement;

  shouldComponentUpdate() {
    return this.props.isComponentUpdated;
  }

  _handleClickHandler = () => {
    this.props.clickHandler();
  };

  componentDidMount () {
    const el = this.node;
    const heightAllow = el.getBoundingClientRect().height + 10;
    const windowHeight = window.innerHeight || document.documentElement.clientHeight;

    if(windowHeight > heightAllow) {
      document.body.style.overflow = 'hidden';
    }

    setTimeout(function(){
      const windowTopPos = window.pageYOffset || document.documentElement.scrollTop;
      const docHeight = document.body.clientHeight;
      const scrollableHeight = (docHeight - heightAllow) > 0 ? (docHeight - heightAllow) : 0;
      if(windowTopPos < scrollableHeight) {
        let allowedHeight = windowHeight - heightAllow;
        el.style.top = (windowTopPos + (allowedHeight>0 ? allowedHeight/2 : 0)) + "px";
      }
      else {
        window.scrollTo( 0, scrollableHeight);
        el.style.top = "";
        el.style.bottom = "0";
      }
      el.style.minHeight = heightAllow + 'px';
      el.style.left = "50%";
    },
    Overlay.leaveTime);
  }


  componentWillUnmount () {
    document.body.style.overflow = '';
  }

  createKeyFeatures = () => {
    const {nature} = this.props.items[this.props.selectedIdx];

    return nature.map(
      (natureItem) => {
        return <div key={natureItem}><i className="fa fa-check-circle-o" aria-hidden="true"></i> {natureItem}</div>
      }
    );

  }

  render() {
    const {animLeft, animTop, items, selectedIdx} = this.props;
    const {title, imgSrc, nature, key, desc} = items[selectedIdx];
    const animateStartPos = (animLeft && animTop) ? {top: animTop, left:animLeft} : {};
    return (
      <div ref={node => this.node=node} className={styles.overlaycard} style={animateStartPos}>
        <div className={styles['overlaycard-bg']}>
          <OverlayIndicator total={items.length} selection={selectedIdx}/>
        </div>
        <div className={styles['overlaycard-image']}>
          <div className={styles['overlaycard-image-adj']}>
            <OverlayCardImageAnimated imgSrc={imgSrc}/>
          </div>
        </div>
        <div className={styles['overlaycard-placement']}>
          <div className={styles['overlaycard-content']}>
            <h2>{title}</h2>
            <div className={styles['overlaycard-content-features']}>
              {this.createKeyFeatures()}
            </div>
            <hr/>
            <div className={styles['overlaycard-content-keyfeature']}>
              <h4>Key Features</h4>
              <span>{key}</span>
            </div>
            <hr/>
            <div className={styles['overlaycard-content-desc']}>
              <h4>Description</h4>
              <span>{desc}</span>
            </div>
            <hr/>
          </div>
          <div className={styles['overlaycard-button']}>
            <Button icon='close' floating accent mini onClick={this._handleClickHandler}/>
          </div>
        </div>
      </div>
    );
  }
}

class OverlayIndicator extends React.Component<OverlayIndicatorProps, {}> {

  static imageAngleStart = 90;

  render() {
    const {total, selection} = this.props;
    const adjTotal = total - 1;
    const angle = Math.floor(180 / adjTotal * selection);
    const adjAngle = angle - OverlayIndicator.imageAngleStart;
    const degInRad = angle * (Math.PI / 180.0);
    const translateY = (Math.sin(degInRad) * 140 * -1) - 20;
    const translateX = ((Math.cos(degInRad) * 140) - 140) * -1;
    const rotation = {transform: 'translate(' +translateX + 'px,' + translateY + 'px) rotate(' + adjAngle + 'deg)'};

    return (
      <div className={styles['overlaycard-bg-big']}>
        <div className={styles['overlaycard-bg-indicator']} style={rotation}>
        </div>
      </div>
    )
  }
}

class OverlayCardImageAnimated extends React.Component<OverlayCardAnimatedProps, {}> {
  private animation = "jello";
  private node:HTMLElement;

  componentDidMount () {
    const el = this.node;
    setTimeout(function(){
      el.classList.add(this.animation);
    },
    Overlay.leaveTime);
  }

  render() {
    const {imgSrc} = this.props;

    return (
      <img src={imgSrc} className="animated" ref={node => this.node=node}/>
    )
  }
}

class Moziac extends React.Component<MoziacProps, {}> {
  private node:HTMLElement;

  _handleClickHandler = () => {
    const el = this.node;
    const posTop = el.getBoundingClientRect().top;
    const posLeft = this.adjustedLeft(el.getBoundingClientRect().left);
    this.props.clickHandler(posTop, posLeft, this.props.idx);
  };

  adjustedLeft = (posLeft:number) => {
    const windowWidth = window.innerWidth || document.documentElement.clientWidth;
    const adjWindowWidth = Math.floor(windowWidth/2);
    if(posLeft > adjWindowWidth) {
      return adjWindowWidth;
    }
    return posLeft;
  };

  render() {
    const {title, imgSrc} = this.props;
    return (
      <div ref={node => this.node=node} className={styles.moziac} onClick={this._handleClickHandler}>
        <img className={styles['moziac-image']} src={imgSrc}></img>
        <div className={styles['moziac-title']}>{title}</div>
      </div>
    );
  }
}
