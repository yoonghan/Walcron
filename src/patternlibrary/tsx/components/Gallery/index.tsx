`use strict`

/**
* Rotating image background for Overlay Indicator is saved as based 64 in scss
* Update to remove props drilling, by replacing Context
**/
import * as React from "react";
import {Transition} from 'react-transition-group';

import {Moziac} from './Moziac';
import {Item, LEAVE_TIME} from './Const';
import {Overlay} from './Overlay';

import '../../../scss/base.scss';

var styles = require('../../../scss/components/Gallery.scss');
declare function require(path: string): any;

interface GalleryState {
  isOverlayOpen: boolean;
  clickPosLeft: number;
  clickPosTop: number;
  selectedIdx: number;
}

export interface GalleryProps {
  items: Array<Item>;
}

export class Gallery extends React.PureComponent<GalleryProps, GalleryState> {

  private transitionStyles:any = {
    entering: {opacity:0},
    entered: {opacity:1},
    exiting: {opacity: 0, transition: 'opacity ' + LEAVE_TIME + 'ms ease-in'}
  };

  constructor(props:any) {
    super(props);
    this.state = {
      isOverlayOpen: false,
      clickPosLeft: 0,
      clickPosTop: 0,
      selectedIdx: 0
    };
  };

  createMoziacListing = ():JSX.Element[] => {
    const moziacListing = this.props.items;

    let idx = 0;
    return moziacListing.map(
      (moziac) =>{
        return (
          <Moziac clickHandler={this.toggleOverlay} imgSrc={moziac.imgSrc} title={moziac.title} key={moziac.title} idx={idx++}></Moziac>
        );
      }
    );
  }

  toggleOverlay = (posTop:number = 0, posLeft:number = 0, selectedIdx:number = 0) => {
    const {isOverlayOpen} = this.state;

    this.setState(
      (prevState, props) => {
        return {
          isOverlayOpen: !isOverlayOpen,
          clickPosLeft: posLeft,
          clickPosTop: posTop,
          selectedIdx: selectedIdx
        };
      }
    );
  };

  render() {
    return (
      <div className={styles.gallery}>
        <Transition in={this.state.isOverlayOpen} timeout={LEAVE_TIME} appear={this.state.isOverlayOpen}>
        {
          (state:any) =>
            {
              if(state === 'exited') {
                return null;
              }

              return (
                  <Overlay
                    transitionStyles={...this.transitionStyles[state]}
                    clickHandler={this.toggleOverlay}
                    animTop={this.state.clickPosTop}
                    animLeft={this.state.clickPosLeft}
                    items={this.props.items}
                    selectedIdx={this.state.selectedIdx}
                    isSubComponentToUpdate={state==='entering'}/>
              );
            }
        }
        </Transition>
        {this.createMoziacListing()}
      </div>
    );
  }
}
