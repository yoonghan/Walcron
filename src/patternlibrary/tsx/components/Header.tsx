`use strict`

import * as React from "react";

import { MenuDrop, MenuDropListing, MenuDropProps } from "./MenuDrop";
import { UtilLocale } from "../util/UtilLocale";
import '../../scss/base.scss';

var styles = require('../../scss/components/Header.scss');
declare function require(path: string): any;

export interface HeaderProp {
  headerTitle: string;
  isHomepage?: Boolean;
  mobileOpensOnLoad?: Boolean;
  menus: Array<MenuDropListing>;
}

export class Header extends React.Component<HeaderProp, {}> {

  constructor(props:any) {
    super(props);
  }

  render() {
    const mnuDropListing:Array<MenuDropListing> = this.props.menus;
    const {headerTitle, isHomepage} = this.props;
    return (
      <div className={styles.hdr}>
        <div className={styles['hdr-div']}>
          <h4 className={styles['hdr-remark']}>
            "{headerTitle}"
          </h4>
          <MenuDrop listing={mnuDropListing} mobileOpensOnLoad={this.props.mobileOpensOnLoad}/>
          <div className={styles['hdr-logo']}>
            <a href={UtilLocale.getLocalizedHref('/')}>
              {isHomepage && <img src='/ext/img/logo/logoOnlyWhiteBg.svg' />}
              {!isHomepage && <img src='/ext/img/logo/logoAsBackButton.svg' />}
            </a>
          </div>
        </div>
      </div>
    );
  }
}
