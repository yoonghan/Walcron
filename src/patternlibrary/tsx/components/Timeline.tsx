`use strict`

import * as React from "react";

import '../../scss/base';
var styles = require('../../scss/components/Timeline');
declare function require(path: string): any;

type TimelineDisplayProp = YearOnlyDisplay | TimelineContainerProp;

export interface TimelineProp {
  postArray?: Array<TimelineDisplayProp>;
}

export interface TimelineContainerProp {
  title: string;
  image?: string;
  description: string;
  linkArray?: Array<Link>;
}

export interface YearOnlyDisplay {
  year: number;
}

export interface Link{
  text: string;
  path: string;
}


//Typescript syntax.
function isTimelineContainer(timelineContainerProp: any): timelineContainerProp is TimelineContainerProp {
  return timelineContainerProp.title;
}

export class Timeline extends React.Component<TimelineProp, {}> {
  constructor(props:any) {
    super(props);
  }

  _createTimelineObject = (post: TimelineContainerProp) => {
    return [
      (<div key={0} className={styles['box-container']}>
        <div className={styles['box']}>
          <TimelineContainer {...post}/>
        </div>
      </div>),
      (<div key={1} className={styles['icon']}></div>)
    ];
  }

  _createYearTimelineObject = (post: YearOnlyDisplay) => {
    return [
      (<div key={0} className={styles['year-wrapper']}></div>),
      (<div key={1} className={styles['year']}><h5>{post.year}</h5></div>)
    ];
  }

  _isEven = (counter: number):Boolean => {
    return (counter % 2 == 0);
  }

  _createTimelinePost = ():JSX.Element[] => {
    const postListing = this.props.postArray;
    const self = this;

    let counter = 0;
    let timelineCounter = 1;

    return postListing.map(
      (post: TimelineDisplayProp) => {
        counter++;
        const aTimelineContainer = isTimelineContainer(post);
        timelineCounter = aTimelineContainer ? (1 + timelineCounter) : timelineCounter;
        //ommit using css to control even, best way may be .primary + .primary
        return (
          <li key={counter} className={(self._isEven(timelineCounter)? styles['even'] : '')}>
            {aTimelineContainer && self._createTimelineObject(post as TimelineContainerProp)}
            {!aTimelineContainer && self._createYearTimelineObject(post as YearOnlyDisplay)}
          </li>
        );
    });
  }

  render() {
    const postListing = this._createTimelinePost();

    return (
      <div className={styles.timeline}>
        <ul>
          {postListing}
        </ul>
      </div>
    );
  }
}

class TimelineContainer extends React.Component<TimelineContainerProp,{}> {
  constructor(props:any) {
    super(props);
  }

  _createLinks = ():JSX.Element[] => {
    const linkListing = this.props.linkArray;

    let counter = 0;
    return linkListing.map(
      (link) => {
        counter++;
        //const location = link.path.indexOf('http') === 0 ? "_blank": "_self";

        return (
          <a href={link.path} target="_blank" key={counter}>&gt; {link.text}</a>
        );
    });
  }

  render() {
    const {title, image, description, linkArray} = this.props;
    const linkListing = (linkArray && this._createLinks()) || <div className={styles['empty-content']}></div>;

    return (
      <div className={styles['content-container']}>
        <h4>{title}</h4>
        <div className={styles['divider']}/>
        <div>
          {image && <img src={image}/>}
          <div className={styles['content-desc-container']}>{description}</div>
          {linkListing}
        </div>
      </div>
    )
  }
}
