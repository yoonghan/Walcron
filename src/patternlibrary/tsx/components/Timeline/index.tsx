`use strict`

/**
 * Upgraded to use React.Fragment to return multiple elements.
 **/
import * as React from "react";

import {TimelineContainer, TimelineDisplayProp, TimelineContainerProp, YearOnlyDisplay} from './TimelineContainer';

import '../../../scss/base.scss';
var styles = require('../../../scss/components/Timeline.scss');
declare function require(path: string): any;

//Typescript syntax.
function isTimelineContainer(timelineContainerProp: any): timelineContainerProp is TimelineContainerProp {
  return timelineContainerProp.title;
}

export interface TimelineProp {
  postArray?: Array<TimelineDisplayProp>;
}

export class Timeline extends React.PureComponent<TimelineProp, {}> {
  constructor(props:any) {
    super(props);
  }

  _createTimelineObject = (post: TimelineContainerProp) => {
    return(
      <React.Fragment>
        <div key={0} className={styles['box-container']}>
          <div className={styles['box']}>
            <TimelineContainer {...post}/>
          </div>
        </div>
        <div key={1} className={styles['icon']}></div>
      </React.Fragment>
    );
  }

  _createYearTimelineObject = (post: YearOnlyDisplay) => {
    return (
      <React.Fragment>
        <div key={0} className={styles['year-wrapper']}></div>
        <div key={1} className={styles['year']}><h3>{post.year}</h3></div>
      </React.Fragment>
    );
  }

  _isEven = (counter: number):Boolean => {
    return (counter % 2 == 0);
  }

  _createTimelinePost = ():JSX.Element[] => {
    const postListing = this.props.postArray;
    const self = this;

    let counter = 0;
    let timelineCounter = 1;

    return postListing.map(
      (post: TimelineDisplayProp) => {
        counter++;
        const aTimelineContainer = isTimelineContainer(post);
        timelineCounter = aTimelineContainer ? (1 + timelineCounter) : timelineCounter;
        //ommit using css to control even, best way may be .primary + .primary
        return (
          <li key={counter} className={(self._isEven(timelineCounter)? styles['even'] : '')}>
            {aTimelineContainer && self._createTimelineObject(post as TimelineContainerProp)}
            {!aTimelineContainer && self._createYearTimelineObject(post as YearOnlyDisplay)}
          </li>
        );
    });
  }

  render() {
    const postListing = this._createTimelinePost();

    return (
      <div className={styles.timeline}>
        <ul>
          {postListing}
        </ul>
      </div>
    );
  }
}
