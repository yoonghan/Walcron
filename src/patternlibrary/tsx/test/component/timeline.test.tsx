import * as React from 'react';
import {Timeline} from "../../components/Timeline";
import * as renderer from 'react-test-renderer';

const timelines = [
  { year: 2016},
  {
    title: "Test",
    image: "https://placeimg.com/640/250/tech",
    altImage: "https://placeimg.com/640/250/tech",
    description: "Description",
    linkArray: [
      {
        text: "text",
        path: "path"
      }
    ]
  },
  { year: 2015},
  {
    title: "Test",
    image: "https://placeimg.com/640/250/tech",
    description: "Description",
    linkArray: [
      {
        text: "text",
        path: "path"
      }
    ]
  },
  {
    title: "Test",
    image: "https://placeimg.com/640/250/tech",
    description: "Description",
    linkArray: [
      {
        text: "text",
        path: "path"
      }
    ]
  }
];

test('Timeline is alternate left right, unaffected by year', () => {
  const component = renderer.create(
    <Timeline postArray = {timelines}/>,
  );
  let tree = component.toJSON();
  expect(tree).toMatchSnapshot();
});
