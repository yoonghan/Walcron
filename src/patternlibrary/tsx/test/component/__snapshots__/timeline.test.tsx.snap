// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Timeline is alternate left right, unaffected by year 1`] = `
<div
  className="timeline"
>
  <ul>
    <li
      className=""
    >
      <div
        className="year-wrapper"
      />
      <div
        className="year"
      >
        <h5>
          2016
        </h5>
      </div>
    </li>
    <li
      className="even"
    >
      <div
        className="box-container"
      >
        <div
          className="box"
        >
          <div
            className="content-container"
          >
            <div />
            <h4>
              Test
            </h4>
            <div
              className="divider"
            />
            <div>
              <img
                className="clickable"
                onClick={[Function]}
                src="https://placeimg.com/640/250/tech"
              />
              <div
                className="content-desc-container"
              >
                Description
              </div>
              <a
                href="path"
                target="_blank"
              >
                &gt; 
                text
              </a>
            </div>
          </div>
        </div>
      </div>
      <div
        className="icon"
      />
    </li>
    <li
      className="even"
    >
      <div
        className="year-wrapper"
      />
      <div
        className="year"
      >
        <h5>
          2015
        </h5>
      </div>
    </li>
    <li
      className=""
    >
      <div
        className="box-container"
      >
        <div
          className="box"
        >
          <div
            className="content-container"
          >
            <h4>
              Test
            </h4>
            <div
              className="divider"
            />
            <div>
              <img
                className={undefined}
                onClick={[Function]}
                src="https://placeimg.com/640/250/tech"
              />
              <div
                className="content-desc-container"
              >
                Description
              </div>
              <a
                href="path"
                target="_blank"
              >
                &gt; 
                text
              </a>
            </div>
          </div>
        </div>
      </div>
      <div
        className="icon"
      />
    </li>
    <li
      className="even"
    >
      <div
        className="box-container"
      >
        <div
          className="box"
        >
          <div
            className="content-container"
          >
            <h4>
              Test
            </h4>
            <div
              className="divider"
            />
            <div>
              <img
                className={undefined}
                onClick={[Function]}
                src="https://placeimg.com/640/250/tech"
              />
              <div
                className="content-desc-container"
              >
                Description
              </div>
              <a
                href="path"
                target="_blank"
              >
                &gt; 
                text
              </a>
            </div>
          </div>
        </div>
      </div>
      <div
        className="icon"
      />
    </li>
  </ul>
</div>
`;
